src/tools/miri/tests/pass/dyn-star-aggregate.stderr

```diff
@@ -0,0 +1,191 @@
+thread 'rustc' panicked at 'assertion failed: match (base.layout.abi, field_layout.abi) {/n    (Abi::Scalar(..), Abi::Scalar(..)) => true,/n    (Abi::ScalarPair(..), Abi::ScalarPair(..)) => true,/n    _ => false,/n}', /home/ec2-user/rust/compiler/rustc_const_eval/src/interpret/projection.rs:LL:CC
```

Comment:
I have no idea what's going on here.

Reply:
It's this assertion

https://github.com/rust-lang/rust/blob/390551716a5f16907e48b21934a66090fd8ef78d/compiler/rustc_const_eval/src/interpret/projection.rs#L109-L113

So looks like a bad layout... strange that that would not be caught by the layout sanity check.

Reply:
What is the `dyn*` layout in that case? It can't be ScalarPair -- if the data part is aggregate, the `dyn*` must also be aggregate.

Reply:
Ah right, but that ScalarPair also doesn't know the actual type of its first field... I don't think I understand why we have a problem here. What is the base and field type in this case where the assertion fails?

`unpack_dyn_star` first projects to the field and then transmutes it... maybe that is the problem? That field could be a `Scalar`, but now has aggregrate layout. But then I would expect it to hit [this assertion](https://github.com/rust-lang/rust/blob/fd1f1fa0d10430885478276341e4afa2e2875adb/compiler/rustc_const_eval/src/interpret/place.rs#L547-L551), not the one in the projection code...

Reply:
> `unpack_dyn_star` first projects to the field and then transmutes it...

Yeah it's definitely this. I added an extra validity assertion in `unpack_dyn_star` and it fails there:

```diff
diff --git a/compiler/rustc_const_eval/src/interpret/place.rs b/compiler/rustc_const_eval/src/interpret/place.rs
index 3c463500a60..323070d4e4a 100644
--- a/compiler/rustc_const_eval/src/interpret/place.rs
+++ b/compiler/rustc_const_eval/src/interpret/place.rs
@@ -859,6 +859,10 @@ pub(super) fn unpack_dyn_star(
         let (ty, _) = self.get_ptr_vtable(vtable)?;
         let layout = self.layout_of(ty)?;
         let data = data.transmute(layout);
+
+        // :(
+        self.validate_operand(&data)?;
+
         Ok((data, vtable))
     }
 }
```

Reply:
Hm, this is nasty. Usually the solution for issues like this is to do `force_allocation`, but then `unpack_dyn_star` requires `&mut self` and we are using it from the visitor which is supposed to work with `&self`...

Reply:
Yeah, I tried that too and ran into exactly the &mut problem you talked about. It *did* work, though, except for the fact that I had to comment out some validation code that only had access to &self 🥲

Reply:
Would be good to get some codegen expertise here. Looks like this PR adds a bunch of ptr casts to codegen to align things; no idea if that breaks some invariants somewhere.

On the Miri side this is pretty fundamental: we could start out with an `Immediate(ScalarPair)` of type `dyn*`, and then after resolving to the actual dynamic type we realize the layout is `Abi::Aggregate` which cannot usually ever be stored in an immediate. This could happen in the middle of the immutable visitor, so we can't force a new allocation either -- we are entirely stuck. Either we somehow make Miri work with `Immediate::Scalar` that do not have `Abi::Scalar` (:scream:) or `dyn*` has to be `Aggregate` from the start (which also affects codegen so it doesn't seem great that Miri would force such a choice).

Cc @oli-obk @eddyb

Reply:
Well I guess another option would be to not allow `dyn*` for types with `Abi::Aggregate`.

Reply:
I'm happy to further constrain the `PointerLike` trait (enforcing `dyn*` coerceability) to require that the type is scalar. In the future, we may want to allow (e.g.) `repr(C)` data types to be coerced into a `dyn*`, but for now I'm not convinced it requires the additional complications that Ralf elaborated above.